includes/Modules/Sign_In_With_Google.php

```diff
@@ -44,9 +51,82 @@ final class Sign_In_With_Google extends Module implements Module_With_Assets, Mo
 	 * @since 1.137.0
 	 */
 	public function register() {
+		add_filter( 'wp_login_errors', array( $this, 'handle_google_auth_errors' ) );
+
+		add_action( 'login_form_google_auth', array( $this, 'handle_google_auth' ) );
 		add_action( 'login_form', $this->get_method_proxy( 'render_signin_button' ) );
 	}
 
+	/**
+	 * Intercept the page request to process token ID
+	 * and complete Sign in with Google flow.
+	 *
+	 * @since n.e.x.t
+	 */
+	public function handle_google_auth() {
+		$request_method = $this->context->input()->filter( INPUT_SERVER, 'REQUEST_METHOD' );
+
+		if ( 'POST' !== $request_method ) {
+			return;
+		}
+
+		$csrf_cookie = $this->context->input()->filter( INPUT_COOKIE, 'g_csrf_token' );
+		$csrf_post   = $this->context->input()->filter( INPUT_POST, 'g_csrf_token' );
+
+		if (
+			! $csrf_cookie ||
+			! $csrf_post ||
+			$csrf_cookie !== $csrf_post
+		) {
+			wp_safe_redirect( add_query_arg( 'error', 'google_auth_invalid_g_csrf_token', wp_login_url() ) );
+			exit;
+		}
+
+		$client_id = $this->get_settings()->get()['clientID'];
+		$id_token  = $this->context->input()->filter( INPUT_POST, 'credential' );
+		try {
+			$client  = new Google_Client( array( 'client_id' => $client_id ) );
+			$payload = $client->verifyIdToken( $id_token );
+
+			if ( empty( $payload ) ) {
+				wp_safe_redirect( add_query_arg( 'error', 'google_auth_invalid_request', wp_login_url() ) );
+				exit;
+			}
+
+			// @TODO implement further flow using $payload in #9339.
+
+		} catch ( \Exception $e ) {
+			wp_safe_redirect( add_query_arg( 'error', 'google_auth_invalid_request', wp_login_url() ) );
+			exit;
+		}
+	}
+
+	/**
+	 * Adds custom errors if Google auth flow failed.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param WP_Error $error WP_Error instance.
+	 * @return WP_Error $error WP_Error instance.
+	 */
+	public function handle_google_auth_errors( $error ) {
+		$error_code = $this->context->input()->filter( INPUT_GET, 'error' );
+		if ( ! $error_code ) {
+			return $error;
+		}
+
+		switch ( $error_code ) {
+			case 'google_auth_invalid_request':
+			case 'google_auth_invalid_g_csrf_token':
+				$error->add( 'access', __( 'Sign in with Google failed.', 'google-site-kit' ) );
```

Comment:
Let's use `google_auth` as the error code to make sure we don't mess with error codes from other plugins.
```suggestion
				$error->add( 'google_auth', __( 'Sign in with Google failed.', 'google-site-kit' ) );
```